[WIP] Eddy Motion Correction (EMC)

dPys opened the pull request:
*Per @arokem 's request to PR the WIP, even though tests + docs are not yet included.
*After an initial glance from folks, plan is to close this WIP PR and eventually separate it into several PR's, perhaps per workflow (and associated interfaces/funcs) @oesteban ?
*About 50% of the code is adapted from fmriprep utilities and qsiprep's implementation of 3dShore, & around 50% is new.
*Supports 3dShore EMC for multishell data, but now implements it exclusively in Dipy as opposed to BrainSuite.
*More importantly, can implement Tensor and SparseFascicle model for LOO prediction on single-shell data! 
*Registration routines currently use ANTs, but could probably be adapted to use to Dipy instead. In that case, they would most likely require careful retuning.
*Pictures to come, but initial visual checks seem to indicate at least decent quality performance (perhaps even reasonably good performance) as a starting benchmark.

pep8speaks commented:
Hello @dPys, thank you for submitting the Pull Request!

* In the file [`dmriprep/interfaces/bids.py`](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/bids.py):

> [Line 3:1](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/bids.py#L3): [F401](https://duckduckgo.com/?q=pep8%20F401) 'pathlib.Path' imported but unused
> [Line 139:23](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/bids.py#L139): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`dmriprep/interfaces/images.py`](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py):

> [Line 184:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py#L184): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 99 characters)
> [Line 188:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py#L188): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 99 characters)
> [Line 193:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py#L193): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 99 characters)
> [Line 194:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py#L194): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 99 characters)
> [Line 210:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py#L210): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 99 characters)
> [Line 225:30](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py#L225): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 243:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py#L243): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 99 characters)
> [Line 243:113](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py#L243): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 250:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py#L250): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 99 characters)
> [Line 260:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py#L260): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 277:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/images.py#L277): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)



* In the file [`dmriprep/interfaces/vectors.py`](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/vectors.py):

> [Line 9:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/interfaces/vectors.py#L9): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 99 characters)

* In the file [`dmriprep/utils/bids.py`](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/utils/bids.py):

> [Line 209:16](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/utils/bids.py#L209): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file



* In the file [`dmriprep/utils/vectors.py`](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/utils/vectors.py):

> [Line 483:65](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/utils/vectors.py#L483): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file



* In the file [`dmriprep/workflows/dwi/base.py`](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/base.py):

> [Line 196:41](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/base.py#L196): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`dmriprep/workflows/dwi/hmc.py`](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py):

> [Line 6:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L6): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 99 characters)
> [Line 75:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L75): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 99 characters)
> [Line 78:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L78): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 99 characters)
> [Line 86:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L86): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 99 characters)
> [Line 92:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L92): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 99 characters)
> [Line 93:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L93): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 98:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L98): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 99:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L99): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 99 characters)
> [Line 204:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L204): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 99 characters)
> [Line 206:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L206): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 99 characters)
> [Line 336:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L336): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 99 characters)
> [Line 375:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L375): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 402:100](https://github.com/nipreps/dmriprep/blob/839fe65270672377c8bffcc245edc6a5f62110eb/dmriprep/workflows/dwi/hmc.py#L402): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 99 characters)

To test for issues locally, `pip install flake8` and then run `flake8 dmriprep`.

oesteban commented on dmriprep/config/emc_coarse_Affine.json:
```diff
@@ -0,0 +1,25 @@
+    "transform_parameters": [ [ 0.2 ], [ 0.15 ] ],
+    "convergence_threshold": [ 1e-06, 1e-06 ],
+    "convergence_window_size": [ 20, 20 ],
+    "metric": [ "Mattes", "Mattes" ],
```
Is this choice based on speed?

Wouldn't it make more sense some cross-correlation based measure? MI is just going to exacerbate the problems you mentioned about the different content of each direction (pinging @arokem and @mattcieslak).

dPys replied:
I agree @oesteban -- cross-corr seems like it would be preferable here. Curious what @mattcieslak thinks?

mattcieslak replied:
Remember, the fixed (model-estimated) images here will have the same contrast as the moving images so most metrics would be ok here. MI is much faster than CC, so any benefit of CC would be outweighed by its extra computation time. Also, this is only the coarse iteration.

oesteban replied:
Yup, but MI has much flatter gradient - I'm going to make up a number here, but 1 CC iteration is as effective as 10 MI iterations.

oesteban commented on dmriprep/interfaces/bids.py:
```diff
@@ -0,0 +1,139 @@
+                out_file = out_file.format(extra_value=self.inputs.extra_values[i])
+            self._results['out_file'].append(out_file)
+            self._results['compression'].append(_copy_any(fname, out_file))
+        return runtime
```
We already have a DerivativesDataSink:

https://github.com/nipreps/dmriprep/blob/0a609b4c91c97b6b2bb0fad0f35ce2e7d15f44b3/dmriprep/interfaces/__init__.py#L13-L16

Unless there is some new functionality here, this can go away.

dPys replied:
yep!

oesteban commented on dmriprep/interfaces/images.py:
```diff
@@ -45,31 +51,13 @@ def _run_interface(self, runtime):
 
 class _RescaleB0OutputSpec(TraitedSpec):
-    out_ref = File(exists=True, desc='One average b0 file')
+    out_ref = File(exists=True, desc='One median b0 file')
```
Please avoid changes out of scope. This one should've been picked up during the review of #25 - that said, I think it is better "average" here, as it needs to be informative.

dPys replied:
Sounds good.

oesteban commented on dmriprep/interfaces/images.py:
```diff
@@ -103,43 +91,279 @@ def _run_interface(self, runtime):
+    def _list_outputs(self):
+        outputs = self.output_spec().get()
+        outputs['out_file'] = os.path.abspath(self._gen_filename('out_file'))
+        return outputs
```
Look at that! https://github.com/poldracklab/niworkflows/blob/4e9181fd31f3a1b216b5edcf040bd5533c400e9d/niworkflows/interfaces/ants.py#L11-L77

Please use ImageMath from niworkflows (and PR against niworkflows if some functionality is missing)

dPys replied:
Awesome

dPys replied:
So the reason for redefining it outside of niworkflows is that we need it to be able to take an explicit secondary argument and file. In other words, as defined here, it is actually quite different. But we could monkey-patch it?

josephmje replied:
Yes, we can import the niworkflows version and then add to it similar to the `DerivativesDataSink` and `BIDSDataGrabberOutputSpec` examples here: https://github.com/nipreps/dmriprep/blob/master/dmriprep/interfaces/__init__.py#L13-L20

oesteban commented on dmriprep/interfaces/vectors.py:
```diff
@@ -92,8 +99,57 @@ def _run_interface(self, runtime):
+                   fmt=['%.8f'] * 3 + ['%g'])
+
+        self._results['out_rasb'] = reor_rasb_file
+        return runtime
```
Isn't this part of #49 ?

dPys replied:
Yes, but with added functionality. All future changes for this interface will be redirected to the #49 PR

oesteban commented on dmriprep/utils/bids.py:
```diff
@@ -160,3 +160,50 @@ def validate_input_dir(exec_env, bids_dir, participant_label):
+            else:
+                copyfileobj(f_in, f_out)
+
+    return True
```
I think this comes from nipype, why not just import them?

dPys replied:
Didn't see these functions in nipype, but maybe I missed them somewhere?

oesteban commented on dmriprep/utils/images.py:
```diff
@@ -0,0 +1,113 @@
+import numpy as np
```
This new file makes a lot of sense, especially if we can come up with unit tests for all these helper functions.

Could you send a separate PR to split the currently existing functions (extract_b0 and rescale_b0) out to this file so they don't show up in the diff?

dPys replied:
Yes.

oesteban commented on dmriprep/utils/vectors.py:
```diff
@@ -375,3 +375,109 @@ def bvecs2ras(affine, bvecs, norm=True, bvec_norm_epsilon=0.2):
+    might cause systematic bias in rotationally invariant measures, such as FA
+    and MD, and also cause characteristic biases in tractography, unless the
+    gradient directions are appropriately reoriented to compensate for this
+    effect [Leemans2009]_.
```
Undefined reference

dPys replied:
I can just removed this reference, as b-vector rotation is fairly common practice/knowledge at this point.

oesteban commented on dmriprep/utils/vectors.py:
```diff
@@ -375,3 +375,109 @@ def bvecs2ras(affine, bvecs, norm=True, bvec_norm_epsilon=0.2):
     return rotated_bvecs
+
+
+def reorient_vecs_from_ras_b(rasb_file, affines, b0_threshold=B0_THRESHOLD):
```
I would try to keep the b0 threshold encoded in the vector representation object, meaning, no need to keep dragging it all around. Please remove.

oesteban replied:
Actually, I would incorporate the reorientation into the vectors object. Will comment on this at #49. We definitely should push that one first.

dPys replied:
That makes sense

oesteban commented on dmriprep/utils/vectors.py:
```diff
@@ -375,3 +375,109 @@ def bvecs2ras(affine, bvecs, norm=True, bvec_norm_epsilon=0.2):
+        M[3, 3] = 1
+        M[0:3, 3] = aff_mat['fixed'].T
+        affines_ras.append(np.multiply(M, ras_trans))
+        del M
```
will comment on #49, but one thing is rotating the vectors and the other is reading a transform. This should be done separately.

dPys replied:
In my latest push, I've separated them out into multiple functions.

dPys commented:
Thank you @oesteban for all of your helpful feedback. I will be out of town next week so won't be able to work a ton on this after tomorrow. But will do my best to keep the dev of EMC moving...

In the meantime, @arokem -- anything that we can think of to speed up the way that sfm predicts signal (i.e. from a given vector coordinate) would be helpful. If it can't be the ElasticNet itself, then we might have to get creative to make it viable with LOO here. Please let me know if there's anything that can be done.

More soon.
@dPys

mattcieslak commented:
@arokem @dPys one thing that made shoreline reasonably fast with 3dSHORE is that I used the L2 (with LShore and NShore) fit instead of L1 or elasticnet. This reduces the problem to a couple dot products and still resulted in very good signal prediction. Would this work with a sfm?

arokem commented:
Yes. That's a good idea and my hunch is that it would work well: Ridge SFM!
Not too hard to implement either as the DIPY SFM object can take a solver
as an input.

On Fri, Jan 17, 2020 at 4:40 PM Matt Cieslak <notifications@github.com>
wrote:

> @arokem <https://github.com/arokem> @dPys <https://github.com/dPys> one
> thing that made shoreline reasonably fast with 3dSHORE is that I used the
> L2 (with LShore and NShore) fit instead of L1 or elasticnet. This reduces
> the problem to a couple dot products and still resulted in very good signal
> prediction. Would this work with a sfm?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipreps/dmriprep/pull/62?email_source=notifications&email_token=AAA46NUS7S252VRARTY2ERTQ6JFXBA5CNFSM4KHNRSUKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJJLQ2Q#issuecomment-575846506>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA46NVYCQLNKKKKQFIKZ23Q6JFXBANCNFSM4KHNRSUA>
> .
>

dPys commented:
@arokem @oesteban @mattcieslak -- The EMC WIP should now run from start to finish. I've also gone ahead and replaced every ANTs reg call with a dipy / pure-python equivalent. These new reg routines are based on @arokem 's api.py file, but have been modified to read and write to files from disk. The routines have also been more or less tuned to the original coarse/precise rigid/affine reg parameters used with the ANTs calls. 

Another note-- sfm is now incorporated with a ridge/L2 solver, which made prediction much faster, among other modifications.

With these further upgrades, I'm seeing a dramatic speedup as well for EMC as a whole. 

To get a sense of EMC's effectiveness at this point, here is a before-and-after:
<img width="1433" alt="Screen Shot 2020-02-04 at 11 19 58 AM" src="https://user-images.githubusercontent.com/16432683/73769464-53f0f200-4740-11ea-98e9-ff1098db5347.png">
![Uploading Screen Shot 2020-02-04 at 11.19.45 AM.png…]()


Now, to actually PR this beast, I propose the following six separate PR's (@oesteban -- please confirm):
1. Reorient vectors interface (FYI:  now relies on a method that has been integrated into the DiffusionGradientTable class)
2. Registration routines (currently included as a separate module register.py in both utils and interfaces)
3. Signal prediction interface 
4. Miscellaneous image-handling utilities and interfaces (i.e. utils/images.py and interfaces/images.py) 
5. Motion reports interfaces
6. EMC workflows (i.e. emc.py)

oesteban commented:
Taking up from where @dPys left off:

> 1. Reorient vectors interface (FYI: now relies on a method that has been integrated into the DiffusionGradientTable class)

I think we have pretty much all we need already merged. Will look back at this if something is missing.

> 2. Registration routines (currently included as a separate module register.py in both utils and interfaces)

@arokem is working on this.

> 3. Signal prediction interface

I'll extract the code from this PR and separate on to a new one.

> 4. Miscellaneous image-handling utilities and interfaces (i.e. utils/images.py and interfaces/images.py)

I think this is already done in part of in full by @josephmje. As for 1, will come back to this PR if anything is needed.

> 5. Motion reports interfaces + figures/viz

Let's leave this out for now, this is being worked out on a different repo.

> 6. EMC workflows (i.e. emc.py)

Will extract this after 3 is merged.

oesteban commented:
Okay, pushed Derek's branch (and his commits) to https://github.com/nipreps/dmriprep/tree/feat/wip/hmc-dpys, just to be sure we don't lose any lines.

oesteban commented:
Closing for clarity.
